### PR TITLE
Fix gradient accumulation for variable length sequences

### PR DESCRIPTION
**Pull request by kibitzing**
### Summary

Fixed gradient accumulation to handle variable-length sequences correctly. Previously, gradients were scaled by `grad_accum_steps` assuming equal tokens per batch, causing incorrect weighting when padding varies across micro-batches.
### Changes
* Changed loss reduction from `'mean'` to `'sum'`
* Accumulate `total_loss_sum` and `num_tokens` across micro-batches
* Scale gradients by total token count: `param.grad.div_(num_tokens)`

### Motivation
I saw a report on [X](https://x.com/QGallouedec/status/1978189638004097040) that there is a SFT loss error when gradient accumulation.

### Test
* Tested on A100x8

No gradient accumulation vs grad_accum=4 (used 2 GPU)
Screenshot 2025-10-15 at 6 05 33 PM

### Reference
* https://unsloth.ai/blog/gradient
* https://huggingface.co/blog/gradient_accumulation

### CODE REVIEW

**Comment by karpathy** [comment]
yeah, it's one of those things that is not fully right but probably also doesn't super matter. let me think about whether there is a clean way to do it.

**Comment by kibitzing** [comment]
Thank you for the feedback. I totally get your point about the added complexity. I'm happy to modify if you have a cleaner solution in mind or close it and stick with the original simplicity. Let me know!

**Comment by qgallouedec** [comment]
In transformers, this is roughly how we handle it:

```python
batches = [next(dataloader) for _ in range(gradient_accumulation_steps)]
num_tokens_in_batch = get_num_items_in_batch(batches)
for batch in batches:
    outputs = models(**batch)
    shift_logits = outputs.logits[:, :-1]
    shift_labels = batch["inputs"][:, 1:]
    per_token_loss = cross_entropy(shift_logits, shift_labels)
    per_token_loss[batch["attention_mask"][:, :-1]] = 0.0
    loss = per_token_loss.sum() / num_tokens_in_batch
    loss.backward()
optimizer.step()
```

**Comment by shirinyamani** [comment]
I assume the `num_tokens_in_batch` is the non-padded tokens, right? 

> In transformers, this is roughly how we handle it:
> 
> ```python
> batches = [next(dataloader) for _ in range(gradient_accumulation_steps)
> num_tokens_in_batch = get_num_items_in_batch(batches)
> for batch in batches:
>     outputs = models(**batch)
>     shift_logits = outputs.logits[:, :-1]
>     shift_labels = batch["inputs"][:, 1:]
>     per_token_loss = cross_entropy(shift_logits, shift_labels)
>     per_token_loss[batch["attention_mask"][:, :-1]] = 0.0
>     loss = per_token_loss.sum() / num_tokens_in_batch
>     loss.backward()
> optimizer.step()
> ```

**Comment by qgallouedec** [comment]
> I assume the `num_tokens_in_batch` is the non-padded tokens, right?

exact

**Comment by kibitzing** [comment]
Following the HuggingFace approach suggested by @qgallouedec, here's the updated implementation:

```python
# evaluate the gradient
num_tokens = torch.tensor(0, device=device) # the number of "active" tokens of supervision seen

# CHANGE: Pre-collect micro-batches and count total valid tokens
steps = [next(train_iter) for _ in range(grad_accum_steps)]
num_tokens += sum((targets >= 0).sum() for _, targets in steps)
if ddp:
    dist.all_reduce(num_tokens, op=dist.ReduceOp.SUM) # sum over ranks

if num_tokens.item() == 0:
    print0(f"Warning: the number of valid tokens in train targets is 0 at step {step}, skipping model update")
    model.zero_grad(set_to_none=True)
    continue

for train_inputs, train_targets in steps:
    with autocast_ctx:
        # CHANGE: scale here with pre-computed num_tokens
        loss = model(train_inputs, train_targets, loss_reduction='sum') / num_tokens
    train_loss = loss.detach() # for logging
    loss.backward() # accumulate the gradient
```

<details><summary> Current implementation of this PR </summary>

```python
# evaluate the gradient
total_loss_sum = torch.tensor(0.0, device=device)  # sum of losses
num_tokens = torch.tensor(0, device=device) # the number of "active" tokens of supervision seen
for micro_step in range(grad_accum_steps):
    train_inputs, train_targets = next(train_iter)
    with autocast_ctx:
        loss = model(train_inputs, train_targets, loss_reduction='sum')
    total_loss_sum += loss.detach() # for logging
    loss.backward() # accumulate the gradient
    num_tokens += (train_targets >= 0).sum()
if ddp:
    dist.all_reduce(total_loss_sum, op=dist.ReduceOp.SUM)
    dist.all_reduce(num_tokens, op=dist.ReduceOp.SUM) # sum over ranks

# scale gradients by total number of tokens
num_tokens_item = num_tokens.item()
if num_tokens_item == 0:
    print0(f"Warning: the number of valid tokens in train targets is 0 at step {step}, skipping model update")
    model.zero_grad(set_to_none=True)
    continue

for param in model.parameters():
    if param.grad is not None:
        param.grad.div_(num_tokens_item)
```

</details>

<details><summary> Original code </summary>

```python

# evaluate the gradient
num_tokens = torch.tensor(0, device=device) # the number of "active" tokens of supervision seen
for micro_step in range(grad_accum_steps):
    train_inputs, train_targets = next(train_iter)
    with autocast_ctx:
        loss = model(train_inputs, train_targets)
    train_loss = loss.detach() # for logging
    loss = loss / grad_accum_steps # each .backward() is a grad sum => normalize loss here
    loss.backward() # accumulate the gradient
    num_tokens += (train_targets >= 0).sum()
if ddp:
    dist.all_reduce(num_tokens, op=dist.ReduceOp.SUM) # sum over ranks

```

</details>

Only added `steps = [next(train_iter) for _ in range(grad_accum_steps)]` and moved `all_reduce` step before computing loss and backward to compute total token count early. (and added `num_tokens_item==0` check as well for safety check)

If this looks good to you, I'll go ahead and test the updated implementation. Please let me know if you have any feedback!

**Comment by svlandeg** [comment]
> Following the HuggingFace approach suggested by @qgallouedec, here's the updated implementation:

I implemented and evaluated this suggestion over at https://github.com/karpathy/nanochat/pull/400 for easy comparison of the two approaches.

**Comment by svlandeg** [comment]
As this has become out-dated since the latest refactors of chat_sft.py, I'll go ahead and close this.
